enhancements/cluster-validation-tool.md

```diff
@@ -0,0 +1,521 @@
+spec:                                                            spec:
+  additionalKernelArgs:                                            additionalKernelArgs:
+    - rcupdate.rcu_normal_after_boot=0                               - rcupdate.rcu_normal_after_boot=0
+  cpu:                                                             cpu:
+    isolated: 2-19,22-39                                             isolated: 2-31,34-63
+    reserved: 0-1,20-21                                              reserved: 0-1,32-33
+  hugepages:                                                       hugepages:
+    defaultHugepagesSize: 1G                                         defaultHugepagesSize: 1G
+    pages:                                                           pages:
+      - count: 32                                                      - count: 32
+                                                              >          node: 0
+        size: 1G                                                         size: 1G
+  machineConfigPoolSelector:                                       machineConfigPoolSelector:
+    pools.operator.machineconfiguration.openshift.io/master:         pools.operator.machineconfiguration.openshift.io/master:
+  nodeSelector:                                                    nodeSelector:
+    node-role.kubernetes.io/master: ""                               node-role.kubernetes.io/master: ""
+  numa:                                                            numa:
+    topologyPolicy: restricted                                       topologyPolicy: restricted
+  realTimeKernel:                                                  realTimeKernel:
+    enabled: true                                             |      enabled: false
+
+---
+<next CR>
+...
+
+Summary
+  Missing required CRs
+    someCR
+    anotherCR
+  Unmatched CRs
+    unknownCR
+    anotherUnknownCR
+  Total CRs with drift: 2
+```
+
+#### Tool delivery
+
+The tool will be a standalone binary executable suitable for direct use by customers, developers,
+support engineering, etc. The tool may be installed as a kubectl/oc plugin.
+
+### Reference Configuration Specification
+
+The reference configuration consists of a set of CRs (yaml files with one CR per file) with
+annotations/metadata to describe expected variability. The reference configuration CRs conform to
+these criteria:
+1. Each CR may be annotated as a required (ie a compliant cluster must contain at least one of that
+   CR) or optional CR.
+1. Each CR may contain annotations to indicate fields which have expected user variation or which
+   are optional.
+1. Any field of a reference configuration CR which is not otherwise annotated is required and the
+   value must be as specified in order to be compliant.
+
+#### Reference CR Annotations
+
+Annotations attached to the CRs are used for providing additional data used during validation. In
```

Comment:
I do not like this much, because the annotations make it to the cluster when the validated config is taken as a boilerplate, edited and applied directly.

I would prefer a way that removes the annotations if I ever try simple oc apply -f using the validation pattern. Some other way (side-car files?) would also allow me to quickly create a new validated reference from a cluster dump (by just copying the meta files in).

Now don't take me wrong, it might be the only viable option atm, but I already overloaded by too many annotations when analysing bug reports.

Reply:
The motivating reason for putting the annotation into the CRs was to keep this data as close as possible (ie same file) to the CR itself. This aids in ongoing maintenance of the reference configuration by not having to update/edit additional content beyond the active CRs.

I agree that this annotation is likely to make it to clusters, but I'm not sure that's necessarily a bad thing. For example, when present it becomes a way to identify (programatically) the use case intended for the cluster and can be used to trigger automated rules checking during analysis of support archives. 

If, as noted above, we have a top level yaml which identifies the name and version of the reference configuration, we could consider removing the "part-of" field here (for all CRs it should match the top level name). In that case a reference configuration that was completely missing this annotation (eg initially sourced from a captured dump of a cluster) would default to `component-name: ""` (meaning "baseline") and `required: true` which seem like the correct semantics for that reference.